# Исследование надежности заемщиков


**Цель исследования** <br>
Ответить на следующие вопросы:<br>
- Как влияет семейное положение на факт погашения кредита в срок?
- Как влияет количество детей клиента на факт погашения кредита в срок?
- Как влияет уровень дохода клиента на факт погашения кредита в срок?
- Как разные цели кредита влияют на факт погашения кредита в срок?

**Контекст исследования**<br>
Заказчик — кредитный отдел банка. Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Входные данные**<br> 
Статистика о платёжеспособности клиентов, полученная от банка.

**План исследования**
1. Ознакомление с данными. 
2. Предобработка данных.
3. Проведение первичного анализа, предварительные выводы и гипотезы.
4. Результаты исследования.

</font><font color='Blue'><b>___</b></font>

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

In [22]:
# Для удобства преобразуем используемые методы анализа в функции

def first_analysis(parametr, output_message):  
    print("\033[1mАнализ распределения наблюдений в исходных данных\033[0m")
    for current_var in data[parametr].unique():
        observations = data[data[parametr] == current_var][parametr].count()
        proportion = observations / data[parametr].count()
        if observations < 30:
            color = "\033[31m"
        else:
            color = "\033[0m"
        print(color + f"{output_message} : {current_var}. Количество наблюдений {observations:<5d}, в процентах - {proportion:.2%}")

def second_analysis(category):
    # строим сводную таблицу, в которой указываем общее количество наблюдений, количество просрочек по кредиту и конверсию
    data_temp = target_data.pivot_table(index=category, values='debt', aggfunc=['count', 'sum', 'mean'])
    data_temp.columns = ['Всего кредитополучателей', 'Всего должников', 'Доля должников']
    # Оформим таблицу цветным градиентом, но можно ее вывести и просто display(data_temp)
    display(data_temp.style.format("{:.3f}").background_gradient(cmap='Blues', axis=0))

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Перед началом исследования посмотрим как наблюдения распределены в исследуемых группах.

In [23]:
first_analysis('children', 'Количество детей')

Анализ распределения наблюдений в исходных данных
Количество детей : 1. Количество наблюдений 4808 , в процентах - 22.54%
Количество детей : 0. Количество наблюдений 14091, в процентах - 66.06%
Количество детей : 3. Количество наблюдений 330  , в процентах - 1.55%
Количество детей : 2. Количество наблюдений 2052 , в процентах - 9.62%
Количество детей : 4. Количество наблюдений 41   , в процентах - 0.19%
Количество детей : 5. Количество наблюдений 9    , в процентах - 0.04%


Красным цветом подсвечены группы, в которых наблюдений недостаточно для проведения анализа. В нашем случае это заниматели с 5-ю детьми, так как эти наблюдения не могут достоверно отобразить свойства своей группы, откажемся от них при проведении исследования.

In [24]:
# исключаем из рассмотрения занимателей с 5 детьми
target_data = data[data['children'] != 5]

Исследование будем проводить, опираясь на коэффицент конверсии возврата кредита в срок. Построим сводную таблицу:

In [25]:
# строим сводную таблицу, в которой указываем общее количество наблюдений, количество просрочек по кредиту и конверсию
second_analysis('children')

Рассмотрим средние показатели возврата кредита в срок для заемщиков с детьми и без. 

Проведем категоризацию заемщиков по наличию детей, создадим в датафрейме `data` столбец `children_category` с категориями: </br>
`0` - нет детей</br>
`1` - есть дети

In [26]:
# функция для категоризации
def categorize_children(children):
    try:
        if children == 0:
            return 0
        elif children > 0:
            return 1
    except:
        pass

# категоризуем данные
data['children_category'] = data['children'].apply(categorize_children)

Построим сводную таблицу:

In [27]:
pivot_table = data.groupby('children_category')['debt'].agg(['count', 'sum', 'mean'])
pivot_table['mean'] = (pivot_table['mean'] * 100).round(2)
pivot_table['on_time'] = 100 - pivot_table['mean']
print(pivot_table)

                   count   sum  mean  on_time
children_category                            
0                  14091  1063  7.54    92.46
1                   7240   669  9.24    90.76


**Выводы:** 
- Заемщики, не имеющие детей, в среднем чаще выплачивают кредит в срок - 92.46%, против - 90.76%. Это связано с меньшей финансовой нагрузкой у бездетных заемщиков.
- Заемщики, имеющие 1-2-х детей, допускают примерно одинаковую просрочку платежа по кредиту - в 9.23% и 9.45% случаев соответственно.
- При этом многодетные заемщики, имеющие 3-х детей, в среднем чаще выплачивают кредит в срок - в 91.82% случаев. Это может быть объяснено тем, что многодетные родители имеют различные меры социальной поддержки от государства, что помогает им облегчить финансовую нагрузку.
- Но при большем количестве детей, финансовая нагрузка слишком большая, поэтому заемщики, имеющие  4-х детей, в среднем чаще остальных допускают просрочку платежа по кредиту - 9.76%.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Исследование проводим аналогично предыдущему примеру.

In [28]:
first_analysis('family_status', 'Семейное положение')

Анализ распределения наблюдений в исходных данных
Семейное положение : женат / замужем. Количество наблюдений 12261, в процентах - 57.48%
Семейное положение : гражданский брак. Количество наблюдений 4134 , в процентах - 19.38%
Семейное положение : вдовец / вдова. Количество наблюдений 951  , в процентах - 4.46%
Семейное положение : в разводе. Количество наблюдений 1189 , в процентах - 5.57%
Семейное положение : Не женат / не замужем. Количество наблюдений 2796 , в процентах - 13.11%


Построим сводную таблицу:

In [29]:
second_analysis('family_status')

Проведем категоризацию заемщиков по состоянию в браке, создадим в датафрейме `data` столбец `family_status_category` с категориями:<br>
`1` - был в браке (`вдовец / вдова`, `разводе`, `женат / замужем`)<br>
`0` - не был в браке (`гражданский брак`, `Не женат / не замужем`)

In [30]:
# функция для категоризации
def categorize_family_status(family_status):
    try:
        if family_status in ['вдовец / вдова', 'разводе', 'женат / замужем']:
            return 1
        elif family_status in ['гражданский брак', 'Не женат / не замужем']:
            return 0
    except:
        pass

# категоризуем данные
target_data['family_status_category'] = target_data['family_status'].apply(categorize_family_status)

C:\Users\marid\AppData\Local\Temp\ipykernel_20896\4275580024.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  target_data['family_status_category'] = target_data['family_status'].apply(categorize_family_status)


Построим сводную таблицу:

In [31]:
second_analysis('family_status_category')

**Выводы:** 
- Люди, имеющие опыт семейной жизни в браке, чаще выплачивают кредит в срок - в 92.51% случаев, против - 90.51%. Скорее всего на это влияют личностные качества такие, как ответсвенность, и  понимание взаимобязательств, которые соблюдаются в браке.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Напомним, что уровень дохода категоризирован в соответсвии со следующей легендой:
- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

Посмотрим как наблюдения распределены в исследуемых группах:

In [32]:
first_analysis('total_income_category', 'Уровень дохода')

Анализ распределения наблюдений в исходных данных
Уровень дохода : B. Количество наблюдений 5014 , в процентах - 23.51%
Уровень дохода : C. Количество наблюдений 15921, в процентах - 74.64%
Уровень дохода : D. Количество наблюдений 349  , в процентах - 1.64%
Уровень дохода : E. Количество наблюдений 22   , в процентах - 0.10%
Уровень дохода : A. Количество наблюдений 25   , в процентах - 0.12%


Не будем исключать группы с очень низким и очень высоким доходом, продолжим исследование опираясь на эти данные.

Построим сводную таблицу:

In [33]:
second_analysis('total_income_category')

**Выводы:** 
- Заемщики с низким доходом менее 30000 (`E`) чаще остальных имеют просрочку по кредиту - в 9.09% случаев. Отнесем их к ненадежным занимателям.
- Заемщики с доходом ниже среднего 30000–50000 (`D`) чаще остальных возвращают кредит в срок - в 93.98% случаев. Скорее всего это связано с более ответсвенным подходом к деньгам, так как люди не располагающие лишними средствами не захотят оказаться в ситуации, когда нужно будет платить штраф за просроченный платеж.
- Заемщики с доходом выше среднего и высоким (`B` и `A`) являются надежными занимателями, они возвращают кредит в срок - в 92.94% и 92.00% случаев соответсвенно.

**Проведем дополнительное исследование.**

In [35]:
# Разделим данные о доходах на категории с помощью метода `.qcut`
# цикл с перебором количества категорий
target_data = data.copy()
start, stop = 3, 8 # начальное и конечное количество категорий
while start <= stop:
    target_data['new_category'] = pd.qcut(target_data['total_income'], q=start, precision=0)
    second_analysis('new_category')
    start += 1

**Выводы:** 
- Заемщики с высоким доходом, более 210 тыс. рублей, являются надежными занимателями, они возвращают кредит в срок - в 92.99% случаев, что является наибольшим показателем возврата в срок среди остальных групп.
- Самая кредито-опасная категория дохода заемщиков с доходом 120 - 140 тыс. рублей.

</font><font color='Blue'><b>___</b></font>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [36]:
first_analysis('purpose_category', 'Цель')

Анализ распределения наблюдений в исходных данных
Цель : операции с недвижимостью. Количество наблюдений 10751, в процентах - 50.40%
Цель : операции с автомобилем. Количество наблюдений 4279 , в процентах - 20.06%
Цель : получение образования. Количество наблюдений 3988 , в процентах - 18.70%
Цель : проведение свадьбы. Количество наблюдений 2313 , в процентах - 10.84%


In [37]:
second_analysis('purpose_category')

**Выводы:** 
- Люди, которые берут кредит для приобретения недвижимости, чаще остальных выплачивают кредит в срок - в 92.74% случаев. Человек чувствует больше ответсвенности, так как сумма кредита очень большая, соответсвенно и большие штрафы при просрочке платежа.
- Также люди, которые берут кредит для проведения свадьбы скорее выплатят все в срок - в 92.09% случаев, так как при создании семьи финансовая нагрузка по кредиту скорее всего будет разделена между супругами.
- Кредит на образование или операции с автомобилем допускают большую просрочку платежа - в 9.25% и 9.35% случаев.

<span style="color:blue">**Замечание**</span>

Выводы, сделанные в этом исследовании, можно рассматривать как *гипотезы* для дальшего статистичекого анализа. Так как нельзя оценить зависимость (корреляцию) между данными, опираясь на методы изученные в 1-2 спринте.

При работе с данными выявленна следующая проблема: при разбиении на группы наблюдения распределены неравномерно, например, размер наибольшей и наименьшей группы отличается в 56 раз! Для дальнешего анализа данные нужно разделить на равные выборки. Также в некоторых группах, слишком мало наблюдений, поэтому выводы сделанные относительно ее могут оказаться недостоверными. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:*

Пропуски были в столбцах общий трудовой стаж (`days_employed`) и ежемесячный доход (`total_income`). Скорее всего пропуски появились из-за того, что некоторые люди решили не разглашать свой доход и трудовой стаж и специально оставили эти сведения пустыми. Но не исключено, что данные были получены объединением наборов данных с разными стандартами, т.е. в одном наборе требование заполнить эти поля было обязательным, а в другом - необязательм.

<font color='DarkMagenta'> Существует несколько причин отсутствия данных в датасетах, в том числе:<br>
- Человеческий фактор: Ошибки ввода данных, не желание отвечать на определенные вопросы.
- Технические проблемы: Возможны сбои с оборудованием или программным обеспечением для сбора данных. Отсутствующие данные могут быть удалены в процессе обработки или очистки данных.
- Организационные: Конфиденциальность, могут отсутствовать данные, идентифицирующие отдельных лиц или компании.</font>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*

**Мода** <br>
Сложность с модой заключается в том, у нас может быть распределение, в котором одно из крайних значений встречается больше раз, чем остальные, поэтому мода не будет отражать распределения всех значений.
Поэтому *моду лучше использовать для категориальных переменных*. Например, мы можем получить моду для пола или города.

**Среднее арифметическое**<br>
Среднее арифметическое сильно зависит от того, насколько максимальные и минимальные значения отличаются от всех остальных значений, т.е. *среднее чувствительно к выбросам*. Влиянию выбросов часто подвержены денежные показатели, например зарплаты, стоимость товаров и услуг. Из-за появления выбросов происходит смещение среднего и истинная картина искажается. 

**Медиана** <br>
*Медиана не подвержена действию выбросов*. Поэтому чтобы исключить влияние выбросов следует выбирать медиану.

**Пример.** Рассмотрим среднее и медиану для колонки трудовой стаж (`days_employed`) 

In [38]:
a, b, = data.days_employed.mean(), data.days_employed.median()
print('Среднее', a)
print('Медиана', b)
print('Во сколько раз среднее превышает медиану', round(a/b))

Среднее 67148.97213258488
Медиана 1996.732855437927
Во сколько раз среднее превышает медиану 34


Можем наблюдать очень большой разброс минимального и максимального значения. Из-за появления выбросов произошло смещение среднего в большую сторону, что неверно отражает реальную ситуацию. В таком случае лучше использовать медиану, а не среднее.
<br></font><font color='Blue'><b>___</b></font>

### Шаг 4: общий вывод.

#### Описание данных. 

Данные представлены таблицой из 12 переменных и 21525 наблюдений.

**Описание данных:**<br>
`children` — int64 — количество детей в семье<br>
`days_employed` — float64 — общий трудовой стаж в днях<br>
`dob_years` — int64 — возраст клиента в годах<br>
`education` — object — уровень образования клиента<br>
`education_id` — int64 — идентификатор уровня образования<br>
`family_status` — object — семейное положение<br>
`family_status_id` — int64 — идентификатор семейного положения<br>
`gender` — object — пол клиента<br>
`income_type` — object — тип занятости<br>
`debt` — int64 — имел ли задолженность по возврату кредитов<br>
`total_income` — float64 — ежемесячный доход<br>
`purpose` — object — цель получения кредита

####  Предобработка данных: порядок обработки аномалий, пропусков и дубликатов.

**Пропущенные значения** были выявлены в двух столбцах:<br>
`total_income` — хранит данные о доходах, так как сумму дохода сильнее всего влияет тип занятости, пропуски в этом столбце заполнены медианным значением по каждому типу из столбца `income_type`;<br>
`days_employed`— хранит данные о  трудовом стаже, пропуски в этом столбце заполнены медианными значениями по каждому типу занятости `income_type` (предварительно в этом столбце были обработаны аномалии).

**Аномалии** в данных были выявлены в двух столбцах:<br>
`days_employed` — хранит данные о  трудовом стаже, артефактом в этом столбце были отрицательные значения, обработка заключалась в замене всех отрицательных значений положительными с помощью метода `abs()`.
`children` — хранит данные о  количестве детей клиента, были выявлены два аномальных значения: -1 и 20. Из датафрейма были удалены строки, в которых встречаются такие аномальные значения.

**Дубликаты** выявлены в одном столбце:<br>
`education` - хранит данные об образовании клиента, в этом столбце были обнаружены неявные дубликаты, одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв, все значения были приведены к нижнему регистру.

**Изменение типов данных:**<br>
`total_income`- хранит данные о доходе клиента, вещественный тип данных в столбце заменен на целочисленный с помощью метода `astype()`.

#### Проведенный первичный анализ.

Анализ проводился с использованием сводных таблиц и коэффициента конверсии возврата кредита в срок. Данные разделялись на исследуемые группы, по которым строилась сводная таблица. На основе этого исследования выдвинуты предположения (гипотезы) для дальнейшего статистического анализа.

При работе с данными выявленна следующая проблема: при разбиении на группы наблюдения распределены неравномерно, например, размер наибольшей и наименьшей группы отличается в 56 раз! Для дальнешего анализа данные нужно разделить на равные выборки. Но в некоторых выборках, слишком мало наблюдений, поэтому выводы сделанные относительно ее могут оказаться недостоверными. Рукомендуется запросить у заказчика дополнительные данные для улучшения качества исследования.

#### Результаты исследования влияния указанных факторов на погашение кредита в срок.

**Количество детей:**<br>
Клиенты, не имеющие детей, в среднем чаще выплачивают кредит в срок - 92.46%, против - 90.76%. Это связано с меньшей финансовой нагрузкой у клиентов не имеющих детей.

**Семейное положение:**<br>
Клиенты, имеющие опыт семейной жизни в браке, чаще выплачивают кредит в срок - в 92.51% случаев, против - 90.51%. Оформленные семейные отношения или их прошлый опыт налагают определенную ответственность и приучают к финансовой дисциплине.

**Уровень дохода:**<br>
Клиенты с высоким доходом, более 210 тыс. рублей, являются надежными занимателями, они возвращают кредит в срок - в 92.99% случаев, что является наибольшим показателем возврата в срок среди остальных групп.

**Цель кредита:**<br>
Люди, которые берут кредит для приобретения недвижимости, чаще остальных выплачивают кредит в срок - в 92.74% случаев. Человек чувствует больше ответсвенности, так как сумма кредита очень большая, соответсвенно и большие штрафы при просрочке платежа.<br>
Также люди, которые берут кредит для проведения свадьбы скорее выплатят все в срок - в 92.09% случаев, так как при создании семьи финансовая нагрузка по кредиту скорее всего будет разделена между супругами.<br>
Кредит на образование или операции с автомобилем допускают большую просрочку платежа и выплачиваются вовремя - в 90.75% и 90.65% случаев.  Это связано с тем, что кредиты на автомобиль и обучение несут дополнительные риски связанные или с возможной аварией, или с проблемами трудоустройства после обучения.

#### Портрет клиента

Анализируя результаты исследования составим портреты клиентов:

**Надежный заемщик:** человек когда-либо состояший в браке, детей не имеет, с доходом выше среднего - более 210тыс. рублей. Хочет получить кредит для операций с недвижимостью.

**Ненадежный заемщик:** человек не состояший в браке, имеет детей, с уровнем дохода ниже среднего - менее 100тыс. рублей. Хочет получить кредит для операций с автомобилем.